[core] Use React 18's JSX runtime for v5.x UMD builds

DiegoAndai opened the pull request:
After https://github.com/mui/material-ui/pull/44720, we forgot to keep using React 18 for the UMD builds. React 19 removed their UMD builds, so we need to use React 18 for it.

To fix this, this PR:

- Adds a `react18` dependency to the `mui-material` package, which is an alias for `react@^18.3.1`
- In the rollup config, resolve `jsx/runtime` to the `react18` dependency

To make this work, I had to update:

- `rollup-plugin-commonjs` (deprecated) to `@rollup/plugin-commonjs`
- `rollup-plugin-node-resolve` (deprecated) to `@rollup/plugin-node-resolve`

These are the updated packages that are recommended. Without this update, the fix still works, but with the updated packages, replacing `jsx/runtime` is much, much cleaner.

This also means that `namedExports` is no longer required, as this has been fixed in `@rollup/plugin-commonjs.` I smoke tested with this PR's build and three components:

**Smoke test**: [index.html.zip](https://github.com/user-attachments/files/18234060/index.html.zip)

This PR also fixes the UMD build test.

DiegoAndai commented:
@Janpot I'm curious whether you think this approach might have some unwanted side effects.

aarongarciah reviewed (approved):
Looks good. We can update it later if @Janpot proposes a better solution.

DiegoAndai commented:
@aarongarciah I reworked this as @Janpot and I didn't want to commit to the previous approach 😅 

I updated the description with the new approach, which is much cleaner. I also attached an `index.html` file that uses this PR's UMD build to test it.

Janpot commented on packages/mui-material/scripts/rollup.config.mjs:
```diff
@@ -117,6 +117,13 @@ const nestedFolder = {
     }
 
+    if (importee.indexOf('react/jsx-runtime') === 0) {
+      return path.resolve(
```
would this work with `createRequire(import.meta.url).resolve('react/jsx-runtime')`?

DiegoAndai replied:
@Janpot `createRequire(import.meta.url).resolve('react18/jsx-runtime')` works 👍🏼, updated

Janpot reviewed (approved):
looks great. small suggestion to avoid reaching in the node_modules folder directly
